horovod/torch/adapter.cc

```diff
@@ -0,0 +1,149 @@
+// Copyright 2018 Uber Technologies, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include <TH/TH.h>
+
+#if HAVE_CUDA
+#include <THC/THC.h>
+#endif
+
+#include "adapter.h"
+#include "cuda_util.h"
+#include "tensor_util.h"
+
+namespace horovod {
+namespace torch {
+
+TorchPersistentBuffer::TorchPersistentBuffer(int device, int64_t size)
+    : device_(device) {
+  with_device device_context(device_);
+  if (device_ == CPU_DEVICE_ID) {
+    buffer_ = new char[size];
+  } else {
+#if HAVE_CUDA
+    THCudaCheck(THCudaMalloc(state, (void**)&buffer_, size));
```

Comment:
No destructor to release memory?

Reply:
It was omitted on purpose.  Unfortunately, by the time this destructor would be called in normal circumstances (application shutdown), CUDA context is already destroyed and cudaFree() operations prints nasty errors in the log - in a pretty normal termination scenario.]]

If we add functionality to terminate Horovod without restarting the application, we should revisit this logic.  I'll add a comment with clarifications.